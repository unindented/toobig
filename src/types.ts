import { Chalk } from "chalk";

export interface CommonConfig {
  readonly baselines?: string | Results;
  readonly reporters?: readonly ReporterConfig[];
}

export interface ScanConfig extends CommonConfig {
  readonly budgets: BudgetsConfig;
  readonly cwd?: string;
}

export interface LoadConfig extends CommonConfig {
  readonly results: string | Results;
}

export interface BudgetsConfig {
  readonly [path: string]: string;
}

export type ReporterConfig = string | readonly [string, object];

export interface Results {
  readonly [path: string]: Result;
}

type Writable<T> = {
  -readonly [K in keyof T]: T[K];
};

export type WritableResults = Writable<Results>;

export interface Result {
  readonly path: string;
  readonly size: number;
  readonly maxSize: number;
}

export interface ReturnValue {
  readonly results: Results;
  readonly anyOverBudget: boolean;
}

export type InputStream = NodeJS.ReadableStream;

export type OutputStream = NodeJS.WritableStream;

export interface ReporterConstructor {
  new (options?: object): Reporter;
}

export interface Reporter {
  onRunStart(): Promise<void> | void;
  onResult(result: Result, baseline?: Result): Promise<void> | void;
  onRunComplete(results: Results, baselines?: Results): Promise<void> | void;
}

export interface OutputContext {
  readonly colors: Chalk;
  readonly colorOverride?: Chalk;
  readonly maxPathLength: number;
  readonly verbose: boolean;
}

export interface ResultsContext {
  readonly results: Results;
  readonly baselines?: Results;
  readonly outputContext: OutputContext;
}

export type ResultsWithBaselinesContext = Required<ResultsContext>;

export interface ResultContext {
  readonly result: Result;
  readonly baseline?: Result;
  readonly outputContext: OutputContext;
}

export type ResultWithBaselineContext = Required<ResultContext>;
