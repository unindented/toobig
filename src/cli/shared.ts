import { cosmiconfigSync } from "cosmiconfig";
import { existsSync, lstatSync } from "fs-extra";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const pkg = require("../../package.json") as { name: string };

export const parserConfiguration = {
  "boolean-negation": true,
  "camel-case-expansion": false,
  "combine-arrays": false,
  "dot-notation": false,
  "duplicate-arguments-array": true,
  "flatten-duplicate-arrays": false,
  "halt-at-non-option": false,
  "negation-prefix": "no-",
  "parse-numbers": false,
  "populate--": false,
  "set-placeholder-key": false,
  "short-option-groups": false,
  "strip-aliased": false,
  "strip-dashed": false,
  "unknown-options-as-args": false,
};

export const configParser = (configPath?: string): object => {
  const explorer = cosmiconfigSync(pkg.name);
  const result = isFile(configPath)
    ? explorer.load(configPath)
    : explorer.search(configPath);

  return result != null && typeof result.config === "object"
    ? (result.config as object)
    : {};
};

const isFile = (path?: string): path is string =>
  !!(path && existsSync(path) && lstatSync(path).isFile());
