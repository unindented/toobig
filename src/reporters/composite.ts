import { Reporter, Result } from "../types";

export interface CompositeReporterOptions {
  reporters: Reporter[];
}

export default class CompositeReporter implements Reporter {
  private readonly reporters: CompositeReporterOptions["reporters"];

  public constructor({ reporters }: CompositeReporterOptions) {
    this.reporters = reporters;
  }

  public async onRunStart(): Promise<void> {
    await Promise.all(this.reporters.map((reporter) => reporter.onRunStart()));
  }

  public async onResult(result: Result): Promise<void> {
    await Promise.all(
      this.reporters.map((reporter) => reporter.onResult(result))
    );
  }

  public async onRunComplete(results: Result[]): Promise<void> {
    await Promise.all(
      this.reporters.map((reporter) => reporter.onRunComplete(results))
    );
  }
}
