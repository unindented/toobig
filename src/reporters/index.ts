export { default as CompositeReporter } from "./composite";
export { default as DefaultReporter } from "./default";
export { default as JSONReporter } from "./json";
export { default as JUnitReporter } from "./junit";
export { default as LineReporter } from "./line";
export { default as SummaryReporter } from "./summary";
export { default as TableReporter } from "./table";
export { default as TAPReporter } from "./tap";
