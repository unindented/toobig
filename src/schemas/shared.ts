import Joi from "@hapi/joi";

export const reportersConfigSchema = Joi.array().items(
  Joi.alternatives().try(
    Joi.string(),
    Joi.array().items(Joi.string().required(), Joi.object().required())
  )
);

export const resultsConfigSchema = Joi.array().items(
  Joi.object({
    path: Joi.string().required(),
    size: Joi.number().required(),
    maxSize: Joi.number().required(),
  }).required()
);
