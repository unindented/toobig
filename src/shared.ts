import { parse } from "url";

import axios from "axios";
import {
  Instance as ChalkInstance,
  supportsColor as globalSupportsColor,
} from "chalk";
import { createReadStream, createWriteStream, ensureFileSync } from "fs-extra";

import {
  InputStream,
  OutputContext,
  OutputStream,
  Result,
  ResultContext,
  Results,
  ResultsContext,
  ResultsWithBaselinesContext,
  WritableResults,
} from "./types";

export { globalSupportsColor as supportsColor };

export const isOverBudget = ({ size, maxSize }: Result): boolean =>
  size >= maxSize;

export const getBaselineDifference = ({
  result,
  baseline,
}: Pick<ResultContext, "result" | "baseline">): number =>
  result.size - (baseline ? baseline.size : 0);

export const isUnderBaseline = (baselines?: Results) => (
  result: Result
): boolean => {
  if (baselines === undefined) {
    return false;
  }

  const baseline = baselines[result.path];
  return getBaselineDifference({ result, baseline }) < 0;
};

export const isOverBaseline = (baselines?: Results) => (
  result: Result
): boolean => {
  if (baselines === undefined) {
    return false;
  }

  const baseline = baselines[result.path];
  return getBaselineDifference({ result, baseline }) > 0;
};

export const filterResults = (
  results: Results,
  predicate: (result: Result) => boolean
): Results =>
  Object.keys(results).reduce((acc, key) => {
    if (predicate(results[key])) {
      acc[key] = results[key];
    }
    return acc;
  }, {} as WritableResults);

export const getTotalResultsSize = ({
  results,
}: Pick<ResultsContext, "results">): number => getTotalSize(results);

export const getTotalBaselinesSize = ({
  baselines,
}: Pick<ResultsWithBaselinesContext, "baselines">): number =>
  getTotalSize(baselines);

const getTotalSize = (results: Results): number =>
  Object.values(results).reduce((acc, { size }) => acc + size, 0);

export const getTotalBaselinesDifference = ({
  results,
  baselines,
}: Pick<ResultsWithBaselinesContext, "results" | "baselines">): number =>
  Object.values(results).reduce((acc, { path, size }) => {
    const baseline = baselines[path];
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    return acc + size - (baseline ? baseline.size : 0);
  }, 0);

export const getOutputContext = ({
  supportsColor,
  maxLength = Infinity,
}: {
  supportsColor: boolean;
  maxLength?: number;
}): OutputContext => ({
  colors: new ChalkInstance(!supportsColor ? { level: 0 } : undefined),
  maxLength,
});

export const getOutputStream = (
  output: string | NodeJS.WritableStream
): OutputStream => {
  if (output === "stdout" || output === "stderr") {
    return process[output];
  }

  if (typeof output === "string") {
    ensureFileSync(output);
    return createWriteStream(output);
  }

  return output;
};

export const endOutputStream = (stream: OutputStream): void => {
  if (stream === process.stdout || stream === process.stderr) {
    return;
  }

  stream.end();
};

export const getInputStream = async (
  input: string | NodeJS.ReadableStream
): Promise<InputStream> => {
  if (input === "stdin") {
    return process.stdin;
  }

  if (typeof input === "string") {
    const { protocol } = parse(input);
    return protocol === null
      ? createReadStream(input)
      : createAxiosStream(input);
  }

  return input;
};

const createAxiosStream = async (
  input: string
): Promise<NodeJS.ReadableStream> => {
  const { data: stream } = await axios.get<NodeJS.ReadableStream>(input, {
    responseType: "stream",
    timeout: 10000,
  });
  return stream;
};

export const readInputStream = (stream: InputStream): Promise<string> => {
  return new Promise((resolve, reject) => {
    const contents: string[] = [];

    stream.on("data", (data: string | Buffer) => {
      contents.push(data.toString());
    });

    stream.on("end", () => {
      resolve(contents.join(""));
    });

    stream.on("error", (error: Error) => {
      reject(error);
    });
  });
};
