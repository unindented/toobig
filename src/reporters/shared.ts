import { basename, dirname, relative, resolve, sep } from "path";

import bytes from "bytes";
import terminalLink from "terminal-link";

import { getTotalSize, isOverBudget, supportsColor } from "../shared";
import { OutputContext, Result } from "../types";

export { getTotalSize, isOverBudget, supportsColor };

export const formatSize = ({ size }: Result): string => bytes(size);

export const formatMaxSize = ({ maxSize }: Result): string => bytes(maxSize);

export const formatSizeVsMaxSize = (
  result: Result,
  { colors, colorOverride }: OutputContext
): string => {
  const overMaxSize = isOverBudget(result);
  const color = colorOverride
    ? colorOverride
    : overMaxSize
    ? colors.bold.red
    : colors.dim;
  const operator = overMaxSize ? ">=" : "<";

  return color(`${formatSize(result)} ${operator} ${formatMaxSize(result)}`);
};

export const formatTotalSize = (results: Result[]): string =>
  bytes(getTotalSize(results));

export const formatPath = (
  { path }: Result,
  { colors }: OutputContext
): string => {
  const relativePath = relative(process.cwd(), path);
  const dir = dirname(relativePath);
  const base = basename(relativePath);
  return `${dir !== "." ? colors.dim(dir + sep) : ""}${colors.bold(base)}`;
};

export const linkPath = (result: Result, options: OutputContext): string => {
  const formattedPath = formatPath(result, options);
  return terminalLink(
    formattedPath,
    `file://${resolve(process.cwd(), result.path)}`,
    {
      fallback: () => formattedPath,
    }
  );
};
