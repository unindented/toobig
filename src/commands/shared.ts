import { CompositeReporter } from "../reporters";
import {
  filterResults,
  getInputStream,
  isOverBaseline,
  isOverBudget,
  isUnderBaseline,
  readInputStream,
} from "../shared";
import {
  Reporter,
  ReporterConfig,
  ReporterConstructor,
  Results,
  ReturnValues,
} from "../types";

export const getCompositeReporter = (
  configReporters: readonly ReporterConfig[]
): Reporter => {
  const reporters = configReporters.map((configReporter) => {
    let reporter: Reporter;

    if (typeof configReporter === "string") {
      const ReporterClass = resolveReporter(configReporter);
      reporter = new ReporterClass();
    } else {
      const ReporterClass = resolveReporter(configReporter[0]);
      reporter = new ReporterClass(configReporter[1]);
    }

    return reporter;
  });

  return new CompositeReporter({ reporters });
};

const resolveReporter = (reporterName: string): ReporterConstructor => {
  let module: { default: ReporterConstructor };

  try {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    module = require(`../reporters/${reporterName}`);
  } catch (err) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    module = require(reporterName);
  }

  return module.default;
};

export const loadResults = async ({
  results,
}: {
  results: string | Results;
}): Promise<Results> => {
  if (typeof results !== "string") {
    return results;
  }

  const stream = await getInputStream(results);
  const contents = await readInputStream(stream);
  return JSON.parse(contents) as Results;
};

export const reportResults = async ({
  results,
  baselines,
  reporter,
}: {
  results: Results;
  baselines?: Results;
  reporter: Reporter;
}): Promise<void> => {
  await reporter.onRunStart();

  for (const result of Object.values(results)) {
    await reporter.onResult(
      result,
      baselines ? baselines[result.path] : undefined
    );
  }

  await reporter.onRunComplete(results, baselines);
};

export const getReturnValues = ({
  results,
  baselines,
}: {
  results: Results;
  baselines?: Results;
}): ReturnValues => {
  const resultsUnderBaseline = filterResults(
    results,
    isUnderBaseline(baselines)
  );
  const resultsOverBaseline = filterResults(results, isOverBaseline(baselines));
  const resultsOverBudget = filterResults(results, isOverBudget);

  const anyUnderBaseline = Object.keys(resultsUnderBaseline).length > 0;
  const anyOverBaseline = Object.keys(resultsOverBaseline).length > 0;
  const anyOverBudget = Object.keys(resultsOverBudget).length > 0;

  return {
    results,
    anyUnderBaseline,
    anyOverBaseline,
    anyOverBudget,
  };
};
