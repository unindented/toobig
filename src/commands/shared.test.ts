import { resolve as resolvePath } from "path";

import { CompositeReporter } from "../reporters";
import { Reporter, Results, ReturnValues } from "../types";

import { getCompositeReporter, getReturnValues, reportResults } from "./shared";

describe(".getCompositeReporter", () => {
  it("returns a composite reporter", () => {
    expect(
      getCompositeReporter([
        "line",
        ["table", { template: "markdown" }],
        resolvePath(__dirname, "../reporters/summary"),
      ])
    ).toBeInstanceOf(CompositeReporter);
  });
});

describe(".reportResults", () => {
  const mockResults: Results = {
    "foo/bar.js": { path: "foo/bar.js", size: 4064, maxSize: 4096 },
    "foo/baz.js": { path: "foo/baz.js", size: 4064, maxSize: 4096 },
  };

  let mockReporter: Reporter;

  beforeEach(() => {
    mockReporter = {
      onRunStart: jest.fn().mockName("onRunStart"),
      onResult: jest.fn().mockName("onResult"),
      onRunComplete: jest.fn().mockName("onRunComplete"),
    };
  });

  describe("without baselines", () => {
    beforeEach(async () => {
      await reportResults({ results: mockResults, reporter: mockReporter });
    });

    it("calls `reporter.onRunStart` once", () => {
      expect(mockReporter.onRunStart).toHaveBeenCalledTimes(1);
    });

    it("calls `reporter.onResult` once per result", () => {
      expect(mockReporter.onResult).toHaveBeenCalledTimes(2);
    });

    it("calls `reporter.onRunComplete` once", () => {
      expect(mockReporter.onRunComplete).toHaveBeenCalledWith(
        mockResults,
        undefined
      );
    });
  });

  describe("with baselines", () => {
    const mockBaselines: Results = {
      "foo/bar.js": { path: "foo/bar.js", size: 4032, maxSize: 4096 },
      "foo/baz.js": { path: "foo/baz.js", size: 4064, maxSize: 4096 },
    };

    beforeEach(async () => {
      await reportResults({
        results: mockResults,
        baselines: mockBaselines,
        reporter: mockReporter,
      });
    });

    it("calls `reporter.onRunStart` once", () => {
      expect(mockReporter.onRunStart).toHaveBeenCalledTimes(1);
    });

    it("calls `reporter.onResult` once per result", () => {
      expect(mockReporter.onResult).toHaveBeenCalledTimes(2);
    });

    it("calls `reporter.onRunComplete` once", () => {
      expect(mockReporter.onRunComplete).toHaveBeenCalledWith(
        mockResults,
        mockBaselines
      );
    });
  });
});

describe(".getReturnValues", () => {
  const mockResults: Results = {
    "foo/bar.js": { path: "foo/bar.js", size: 4064, maxSize: 4096 },
    "foo/baz.js": { path: "foo/baz.js", size: 4224, maxSize: 4096 },
    "foo/qux.js": { path: "foo/baz.js", size: 4032, maxSize: 4096 },
  };

  let returnValues: ReturnValues;

  describe("without baselines", () => {
    beforeEach(() => {
      returnValues = getReturnValues({ results: mockResults });
    });

    it("returns the corresponding values", () => {
      expect(returnValues).toMatchSnapshot();
    });
  });

  describe("with baselines", () => {
    const mockBaselines: Results = {
      "foo/bar.js": { path: "foo/bar.js", size: 4224, maxSize: 4096 },
      "foo/baz.js": { path: "foo/baz.js", size: 4064, maxSize: 4096 },
      "foo/qux.js": { path: "foo/qux.js", size: 4032, maxSize: 4096 },
    };

    beforeEach(() => {
      returnValues = getReturnValues({
        results: mockResults,
        baselines: mockBaselines,
      });
    });

    it("returns the corresponding values", () => {
      expect(returnValues).toMatchSnapshot();
    });
  });
});
