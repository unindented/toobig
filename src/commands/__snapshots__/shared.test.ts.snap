// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.getReturnValues with baselines returns the corresponding values 1`] = `
Object {
  "anyOverBaseline": true,
  "anyOverBudget": true,
  "anyUnderBaseline": true,
  "results": Object {
    "foo/bar.js": Object {
      "maxSize": 4096,
      "path": "foo/bar.js",
      "size": 4064,
    },
    "foo/baz.js": Object {
      "maxSize": 4096,
      "path": "foo/baz.js",
      "size": 4224,
    },
    "foo/qux.js": Object {
      "maxSize": 4096,
      "path": "foo/baz.js",
      "size": 4032,
    },
  },
}
`;

exports[`.getReturnValues without baselines returns the corresponding values 1`] = `
Object {
  "anyOverBaseline": false,
  "anyOverBudget": true,
  "anyUnderBaseline": false,
  "results": Object {
    "foo/bar.js": Object {
      "maxSize": 4096,
      "path": "foo/bar.js",
      "size": 4064,
    },
    "foo/baz.js": Object {
      "maxSize": 4096,
      "path": "foo/baz.js",
      "size": 4224,
    },
    "foo/qux.js": Object {
      "maxSize": 4096,
      "path": "foo/baz.js",
      "size": 4032,
    },
  },
}
`;
